<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R package for fast BED-file manipulation • bedtorch</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="R package for fast BED-file manipulation">
<meta property="og:description" content="The goal of bedtorch is to provide a fast BED file manipulation tool suite native in R.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bedtorch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.12.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/haizi-zh/bedtorch/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="bedtorch" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bedtorch" class="anchor"></a>bedtorch</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p><em>For full documentation, refer to the <a href="https://haizi-zh.github.io/bedtorch/">package documentation site</a>.</em></p>
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p>The goal of bedtorch is to provide a fast BED file manipulation tool suite native in R.</p>
<p>In bioinformatics studies, an important type of jobs is related to BED and BED-like file manipulation. To name a few example:</p>
<ul>
<li>Filtering cfDNA fragments based on length</li>
<li>Identifying genomic regions overlapping with certain genes</li>
<li>Filtering features based on epigenetic profiles such as histone modification levels</li>
<li>Extracting SNVs in a certain genomic interval</li>
</ul>
<p>Many of such tasks can be done by some highly-optimized tools, such as <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a href="https://bedops.readthedocs.io/en/latest/">bedops</a>. However, these are binary tools which only works in shell. Sometimes, especially for developers using R, it’s desirable to invoke these tools in R codes.</p>
<p>One solution is using functions that invokes system commands (<a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/system2"><code>system2</code></a> or <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/system"><code>system</code></a>). <a href="https://cran.r-project.org/web/packages/bedr/vignettes/Using-bedr.html"><code>bedr</code></a>, a popular package, is one good example. It’s a bridge linking between R codes and shell tools. Under the hood, it writes data into the temporary directory, calls certain command line tools such as <code>tabix</code>, <code>bedtools</code>, or <code>bedops</code>, parses the output and re-generate a data frame in R space.</p>
<p>Generally this approach is very helpful. However, it’s worth noting some disadvantages.</p>
<ul>
<li>This workflow requires lots of disk IO, since data frames in R needs to be written to temporary directories before being fed to command line tools.</li>
<li>Command line output is captured and transferred back to R s space. The data transfer is done by stdin/stdout pipes. For R, reading large amount of text content from stdin can be extremely slow. As a result, most of time when large dataset is involved, <code>bedr</code> is hardly usable.</li>
<li>
<code>bedr</code> captures command line output as plain text. Therefore, when the data is transferred back to R, most of the information about data types, index (in the case of <code>data.table</code>), column names, etc., are lost. It’s the package user’s responsibility to make sure data types are correct, column names are restored, column indices have been rebuilt (in the case of <code>data.table</code>, which can be very time consuming). Often this job is tedious, error-prone and introduces very large overheads.</li>
<li>It requires the command line tools to be present in <code>PATH</code>. This depends on the end user’s system environment and the R developer can do nothing about it. It puts additional burden to deployment, especially when the end user is not an IT professional.</li>
</ul>
<p>Another approach is using native R data types to represent BED-like data. R’s native <code>data.frame</code> seems to be a good model, but in many scenarios, it’s performance is poor. More importantly, it lacks many advanced features compared with <code>GenomicRanges</code>, <code>dplyr</code> and <code>data.table</code>.</p>
<p><code>GenomicRanges</code> is widely used in bioinformatics community. Although it provides essential building blocks for many tasks, such as <code>findOverlaps</code>, <code>reduce</code>, etc., it doesn’t provide high-level features as <code>bedtools</code> does. Using <code>findOverlaps</code>, <code>reduce</code>, and other functions provided by <code>GenomicRanges</code>, users may implement operations such as map, merge, unique intersect, but this still requires lots of coding.</p>
<p>It would be desirable to have an R package, which represent BED-like data by <code>GenomicRanges</code> or <code>data.table</code> objects, and provide native support for common BED manipulations, such as merge, intersect, filter, etc.</p>
</div>
<div id="details" class="section level2">
<h2 class="hasAnchor">
<a href="#details" class="anchor"></a>Details</h2>
<p><code>bedtorch</code> is an attempt for this. Under the hood, every BED dataset is represented as a <code>GenomicRanges</code> object (it also can represent the dataset as <code>data.table</code>, if preferred. But this is not recommended). Users can apply various operations on it, such as intersect, merge, etc.</p>
<p>In terms of the core computation, <code>bedtorch</code>’s performance is comparable to <code>bedtools</code>. However, since no disk IO and data conversion is needed, users can directly manipulate the dataset in memory, therefore in practice, via <code>bedtorch</code>, many tasks can be done about one magnitude faster.</p>
<p>Additionally, by linking to <a href="https://github.com/samtools/htslib">htslib</a>, <code>bedtorch</code> can write data frames directly to BGZIP-format files, and optionally create the tabix index. It can also directly load either local or remote BGZIP-format files at any particular genomic regions. This feature is done by htslib, therefore not requiring external command line tools such as <code>tabix</code> or <code>bgzip</code>. In the case of working with remote files, being capable of loading any arbitrary portion of the file will be very convenient.</p>
<p><a href="http://www.htslib.org/doc/bgzip.html">BGZIP</a> is a variant of gzip and is widely used in bioinformatics. It’s compatible with gzip, with an important additional feature: allows a compressed file to be indexed for fast query. However, R is unaware of BGZIP. It considers all <code>.gz</code> files as gzip-compressed, thus doesn’t take advantage of BED index, and cannot output BGZIP files.</p>
<p>For more details, refer to the <a href="https://haizi-zh.github.io/bedtorch/reference/index.html">package reference page</a>.</p>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"haizi-zh/bedtorch"</span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level3">
<h3 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h3>
<p>Here are several basic examples which shows you how to solve a common problem:</p>
<p>Read a BED file from disk:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/haizi-zh/bedtorch">bedtorch</a></span><span class="op">)</span>

<span class="co">## Load BED data</span>
<span class="va">file_path</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example2.bed.gz"</span>, package <span class="op">=</span> <span class="st">"bedtorch"</span><span class="op">)</span>

<span class="va">bedtbl</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">file_path</span>, range <span class="op">=</span> <span class="st">"1:3001-4000"</span><span class="op">)</span>
<span class="va">bedtbl</span>
<span class="co">#&gt; GRanges object with 14 ranges and 2 metadata columns:</span>
<span class="co">#&gt;        seqnames    ranges strand |    score1    score2</span>
<span class="co">#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">#&gt;    [1]        1 2926-3011      * |       106       181</span>
<span class="co">#&gt;    [2]        1 3004-3092      * |        88       193</span>
<span class="co">#&gt;    [3]        1 3092-3193      * |       118       212</span>
<span class="co">#&gt;    [4]        1 3165-3248      * |        94       211</span>
<span class="co">#&gt;    [5]        1 3233-3345      * |       107       205</span>
<span class="co">#&gt;    ...      ...       ...    ... .       ...       ...</span>
<span class="co">#&gt;   [10]        1 3619-3726      * |       108       172</span>
<span class="co">#&gt;   [11]        1 3695-3803      * |       117       214</span>
<span class="co">#&gt;   [12]        1 3777-3867      * |       109       195</span>
<span class="co">#&gt;   [13]        1 3845-3939      * |        92       199</span>
<span class="co">#&gt;   [14]        1 3925-4016      * |        93       197</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
<p>Read a remote BGZIP BED file for a certain region:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load remote BGZIP files with tabix index specified</span>
<span class="co"># Here we need to explicitly indicate `compression` as `bgzip` since we are</span>
<span class="co"># using short URL for `tabix_index`, so that the function cannot guess</span>
<span class="co"># compression type using the URL</span>
<span class="fu"><a href="reference/read_bed.html">read_bed</a></span><span class="op">(</span>
  <span class="st">"https://git.io/JYATB"</span>,
  range <span class="op">=</span> <span class="st">"22:20000001-30000001"</span>,
  tabix_index <span class="op">=</span> <span class="st">"https://git.io/JYAkT"</span>,
  compression <span class="op">=</span> <span class="st">"bgzip"</span>
<span class="op">)</span>
<span class="co">#&gt; GRanges object with 21 ranges and 2 metadata columns:</span>
<span class="co">#&gt;        seqnames            ranges strand |      score comp_method</span>
<span class="co">#&gt;           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt; &lt;character&gt;</span>
<span class="co">#&gt;    [1]       22 20000001-20500000      * | -0.0613535    cofrag/3</span>
<span class="co">#&gt;    [2]       22 20500001-21000000      * | -0.0422752    cofrag/3</span>
<span class="co">#&gt;    [3]       22 21000001-21500000      * |  0.0268175    cofrag/3</span>
<span class="co">#&gt;    [4]       22 21500001-22000000      * |  0.0386308    cofrag/3</span>
<span class="co">#&gt;    [5]       22 22000001-22500000      * |  0.0265141    cofrag/3</span>
<span class="co">#&gt;    ...      ...               ...    ... .        ...         ...</span>
<span class="co">#&gt;   [17]       22 28000001-28500000      * | 0.08472735    cofrag/3</span>
<span class="co">#&gt;   [18]       22 28500001-29000000      * | 0.15800267    cofrag/3</span>
<span class="co">#&gt;   [19]       22 29000001-29500000      * | 0.14646452    cofrag/3</span>
<span class="co">#&gt;   [20]       22 29500001-30000000      * | 0.07898761    cofrag/3</span>
<span class="co">#&gt;   [21]       22 30000001-30500000      * | 0.00434186    cofrag/3</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
<p>Write the previous data table to the temporary directory, and create the tabix index alongside:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/write_bed.html">write_bed</a></span><span class="op">(</span><span class="va">bedtbl</span>,
          file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bed.gz"</span><span class="op">)</span>,
          tabix_index <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Merge intervals, and take the mean of score in each merged group:</p>
<p><a href="https://bedtools.readthedocs.io/en/latest/"><img src="https://bedtools.readthedocs.io/en/latest/_images/merge-glyph.png" alt="Image by bedtools" width="64%"></a></p>
<p><em>Image by <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></em></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">operation</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mean_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>on <span class="op">=</span> <span class="st">"score1"</span>, func <span class="op">=</span> <span class="va">mean</span><span class="op">)</span><span class="op">)</span>
<span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/merge_bed.html">merge_bed</a></span><span class="op">(</span><span class="va">bedtbl</span>, operation <span class="op">=</span> <span class="va">operation</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 1 range and 1 metadata column:</span>
<span class="co">#&gt;       seqnames    ranges strand | mean_score</span>
<span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]        1 2926-4016      * |    99.0714</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre></div>
<p>Find intersections between two datasets:</p>
<p><a href="https://bedtools.readthedocs.io/en/latest/"><img src="https://bedtools.readthedocs.io/en/latest/_images/intersect-glyph.png" alt="Image by bedtools" width="64%"></a></p>
<p><em>Image by <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></em></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file_path1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_merge.bed"</span>, package <span class="op">=</span> <span class="st">"bedtorch"</span><span class="op">)</span>
<span class="va">file_path2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_intersect_y.bed"</span>, package <span class="op">=</span> <span class="st">"bedtorch"</span><span class="op">)</span>

<span class="va">tbl_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">file_path1</span>, genome <span class="op">=</span> <span class="st">"hs37-1kg"</span><span class="op">)</span>
<span class="va">tbl_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">file_path2</span>, genome <span class="op">=</span> <span class="st">"hs37-1kg"</span><span class="op">)</span>
<span class="fu"><a href="reference/intersect_bed.html">intersect_bed</a></span><span class="op">(</span><span class="va">tbl_x</span>, <span class="va">tbl_y</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 12 ranges and 1 metadata column:</span>
<span class="co">#&gt;        seqnames    ranges strand |     score</span>
<span class="co">#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span>
<span class="co">#&gt;    [1]       21     23-25      * |         7</span>
<span class="co">#&gt;    [2]       21     27-30      * |         7</span>
<span class="co">#&gt;    [3]       21     30-35      * |         9</span>
<span class="co">#&gt;    [4]       21     48-49      * |         1</span>
<span class="co">#&gt;    [5]       21     48-50      * |         2</span>
<span class="co">#&gt;    ...      ...       ...    ... .       ...</span>
<span class="co">#&gt;    [8]       22     43-44      * |         9</span>
<span class="co">#&gt;    [9]       22     47-50      * |         0</span>
<span class="co">#&gt;   [10]       22     52-59      * |         9</span>
<span class="co">#&gt;   [11]       22     54-57      * |         5</span>
<span class="co">#&gt;   [12]       22     58-60      * |         9</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 84 sequences from hs37-1kg genome</span></code></pre></div>
<p>Shuffle a BED data table across the genome:</p>
<p><a href="https://bedtools.readthedocs.io/en/latest/"><img src="https://bedtools.readthedocs.io/en/latest/_images/shuffle-glyph.png" alt="Image by bedtools" width="64%"></a></p>
<p><em>Image by <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></em></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/shuffle_bed.html">shuffle_bed</a></span><span class="op">(</span><span class="va">tbl_x</span><span class="op">)</span>
<span class="co">#&gt; GRanges object with 20 ranges and 1 metadata column:</span>
<span class="co">#&gt;        seqnames            ranges strand |     score</span>
<span class="co">#&gt;           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span>
<span class="co">#&gt;    [1]       21   1017987-1017992      * |         9</span>
<span class="co">#&gt;    [2]       21   6511120-6511122      * |         5</span>
<span class="co">#&gt;    [3]       21 10654657-10654661      * |         4</span>
<span class="co">#&gt;    [4]       21 12523054-12523056      * |         7</span>
<span class="co">#&gt;    [5]       21 15435384-15435388      * |         2</span>
<span class="co">#&gt;    ...      ...               ...    ... .       ...</span>
<span class="co">#&gt;   [16]       22 30737152-30737159      * |         9</span>
<span class="co">#&gt;   [17]       22 31398105-31398108      * |         5</span>
<span class="co">#&gt;   [18]       22 33878179-33878183      * |         8</span>
<span class="co">#&gt;   [19]       22 35584543-35584546      * |         0</span>
<span class="co">#&gt;   [20]       22 44203020-44203028      * |         3</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 84 sequences from hs37-1kg genome</span></code></pre></div>
<p>Calculate Jaccard statistics between the two BED data tables:</p>
<p><a href="https://bedtools.readthedocs.io/en/latest/"><img src="https://bedtools.readthedocs.io/en/latest/_images/jaccard-glyph.png" alt="Image by bedtools" width="64%"></a></p>
<p><em>Image by <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></em></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/jaccard_bed.html">jaccard_bed</a></span><span class="op">(</span><span class="va">tbl_x</span>, <span class="va">tbl_y</span><span class="op">)</span>
<span class="co">#&gt;    intersection union   jaccard n_intersections</span>
<span class="co">#&gt; 1:           33    94 0.3510638               8</span></code></pre></div>
</div>
<div id="performance" class="section level3">
<h3 class="hasAnchor">
<a href="#performance" class="anchor"></a>Performance</h3>
<p>The following is the result of a simple benchmark for three common BED manipulations: merge, intersect and map.</p>
<p>Benchmark platform information:</p>
<ul>
<li>OS: macOS Big Sur 11.1</li>
<li>CPU: 2.7 GHz Quad-Core Intel Core i7</li>
<li>Memory: 16 GB LPDDR3</li>
</ul>
<p><img src="https://raw.githubusercontent.com/haizi-zh/bedtorch/main/data-raw/benchmark.png" width="64%/"></p>
<p>From the benchmark result, we can see that for all three tasks, <code>bedtorch</code> uses much less time than <code>bedtools</code>.</p>
<p>Note: this does not mean <code>bedtools</code> is indeed slower than <code>bedtorch</code> at performing the actual computation. Don’t forget <code>bedtools</code> requires large amount of disk IO, while <code>bedtorch</code> does not.</p>
</div>
</div>
<div id="see-also" class="section level2">
<h2 class="hasAnchor">
<a href="#see-also" class="anchor"></a>See also</h2>
<p>Many features are inspired by bedtools. Thus, it’s helpful to get familiar with bedtool’s documentation: <a href="https://bedtools.readthedocs.io/en/latest/index.html" class="uri">https://bedtools.readthedocs.io/en/latest/index.html</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/haizi-zh/bedtorch/">https://​github.com/​haizi-zh/​bedtorch/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Haizi Zheng <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/web/licenses/MIT"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="License: MIT"></a></li>
<li><a href="https://github.com/haizi-zh/bedtorch/actions"><img src="https://github.com/haizi-zh/bedtorch/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/haizi-zh/bedtorch/actions"><img src="https://codecov.io/gh/haizi-zh/bedtorch/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://www.codefactor.io/repository/github/haizi-zh/bedtorch"><img src="https://www.codefactor.io/repository/github/haizi-zh/bedtorch/badge" alt="CodeFactor"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Haizi Zheng.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
